menu v2

mcmatrix opened the pull request:
Happy New Year! @KevinOConnor 
This is a beta preview of V2 menu. 
It is still WIP and needs more testing.
The base functionality seems to be settled,
there may be coming some changes or improvements in the code.

I hope to add more information about the new menu system soon.

KevinOConnor commented:
Happy New Year @mcmatrix!

FYI, I'm a bit backlogged right now.  I've been holding the repo in a "soft freeze" for the last couple of weeks for the holidays.  In the next week or so I plan to start merging in a bunch of PRs that have been queued.  Alas, that means I don't think I'll get a chance to look at this code for a week or so.

Cheers,
-Kevin

schnello commented:
Hello

Any plan for an option to reverse the encoder only for the menu navigation? 
So that the user can decide how the encoder operates separately with regard to values and menu navigation

Christian

mcmatrix commented:
Hi @schnello 

Isnt that already possible by just changing the order of encoder pins in config?
The input value has reverse attribute that can be used.

I can test this tomorrow, to see does the order change have effect.

mcmatrix commented:
Hi @KevinOConnor 
That's fine. Meanwhile, I'll try to continue with testing.
Br,
-Janar

mcmatrix commented:
Hi @schnello 

I added boolean `menu_reverse_navigation` option to the menu but it's still in my work branch.
Sorry, my days have been busy, hope to push this update soon to PR.

Btw if you want to test my work branch then following should help
```
git remote add mcmatrix https://github.com/mcmatrix/klipper
```
```
git fetch mcmatrix
git checkout mcmatrix/work-menu-20191122

make clean
make

sudo service klipper stop
make flash FLASH_DEVICE=/dev/ttyACM0
sudo service klipper start
```

schnello commented:
@mcmatrix 
Thanks a lot. Compiled and installed at SKR Pro 1.1... works fine. Also the "menu_reverse_navigation" option works as expected.

schnello commented:
@mcmatrix 
Testet now some days and it works perfekt.
One point:
```
[menu __control __move_10mm __axis_z]
type: input
enable: {not printer.idle_timeout.state == "Printing"}
name: Move Z:{'%05.1f' % me.input}
input: {printer.toolhead.position.z}
input_min: {printer.menu.default.axis_z_min}
input_max: {printer.menu.default.axis_z_max}
input_step: 10.0
script_shortpress:
    G90
    G1 Z{me.input}

[menu __control __move_10mm __axis_e]
type: input
enable: {not printer.idle_timeout.state == "Printing"}
name: Move E:{'%+06.1f' % me.input}
input: 0
input_min: {printer.menu.default.axis_e_min}
input_max: {printer.menu.default.axis_e_max}
input_step: 10.0
script_shortpress:
    M83
    G1 E{me.input} F240
```

If you move the first time XYZ all is ok. But if you move E the force is set to F240. From now XYZ is also set to F240 and is super slow. So i add in the menu a fixes F Value and the problem is solved.

schnello commented:
@mcmatrix 
further observation:
Go to the Controll Menu and set the FAN Speed to 20%
Then move to FAN off/on and aktive the Menu Entry... the FAN changed now the speed to 100%

KevinOConnor commented:
Thanks and sorry for my delay in responding.

Alas, I fear the commit is too large for me to meaningfully review.  I did look through the interface between the menu code and the rest of the system (specifically the config file and changes to non-menu code).

One thing I saw:
```
script_render:
    {%- set width = 14 -%}
    {%- set centre = (printer.menu.cols - width) // 2 -%}
    `{' '*centre}{'='*width}
    `{' '*centre}<?item:.__offsetz?>
    `{' '*centre}{'='*width}
```
I'm concerned that the menu code is introducing it's own config/programming language (ie, the above is not Python, Jinja2, "ini config" format, or g-code).  My high-level feedback is that I don't think it is a good idea to introduce a new language.  (Specifically, I fear the number of people willing/able to use/maintain a new language is too small.)

Are there practical alternatives to menu configuration that could use basic Jinja2 and ini config file format?  My high-level feedback is that I, at least personally, would rather have a menu system that is simple to document and configure than a full featured menu system.

-Kevin

mcmatrix commented:
Thank you for review.
I won't call it a new language.
This script is pure jinja2 script
```ini
script_render:
    {%- set width = 14 -%}
    {%- set centre = (printer.menu.cols - width) // 2 -%}
    `{' '*centre}{'='*width}
    `{' '*centre}<?item:.__offsetz?>
    `{' '*centre}{'='*width}
```
The placeholder `<?item:.__offsetz?>` will not interact with render, it will be replaced after rendering.

These `<?back:xxxx?>`, `<?item:xxxx?>`, `<?name:xxxx?>` placeholders similar to `~glyph~` are needed to bind menu elements and jinja2 script output. Needed to make menu scrolling and movements work.

I need to get menu item list from script before rendering to build an initial items and replace placeholders after rendering. Otherwise, I don't know the position of that item in the rendered output.

I don't know is there a way to simplify it more without losing functionalities.
What are your exact concerns?

KevinOConnor commented:
> Is the concept more or less okay, beeing both menu and screen, a potential replacement for display, named scripts, fewer menu types etc?

Alas, I'm struggling to review, so it's hard for me to say.  I'd certainly like to see both the display and menus configurable.  I'm not sure what you mean by "named scripts", but I'm not aware of a reason to implement them.

> What parts of this menu v2 are your exact concerns?

I see what appears to be a new syntax for the menu.  Specifically, when I look through the updated menu configuration, I see `{%- ... -%}` (where Klipper uses `{% %}`), I see backticks in script fields, I see `<?item:.__offsetz?>`, I see `blinking_mask: ++++++...+`, I see `me.runtime_names`.  I don't know what these things do and I fear general users and developers will react similarly.  I fear introducing a syntax that enables advanced features risks making it harder to implement basic menu configuration.  Said another way, my preference would be a very simple configuration, even if that means the menus lack advanced features.

FWIW, what I expected to see when looking at the configuration is the ability to define menu lists (choose one of N options - either a submenu or a selection), and to define numeric menu screens (rotate the dial to select a value between min/max values).  I'd expect each menu config section to evaluate most config parameters via Jinja2 (eg, script to run, display name, min/max value).  When I look at the config I don't immediately see these things - from my perspective, I see a "new syntax" that accomplishes goals which I'm not sure need to be accomplished.

-Kevin

mcmatrix commented:
Thank you for the honest review.

One thing I want immediately clarify
> I see {%- ... -%} (where Klipper uses {% %}), 

This is pure Jinja2, it's how jinja is controlling whitespaces.
Excessive newlines are something I want to avoid.
The other option is to enable jinja `trim_blocks` and `lstrip_blocks` parameters.
https://jinja.palletsprojects.com/en/2.11.x/templates/#whitespace-control

I agree that it's not the simplest functionality and configuration.
I have your expectations about the menu system. I'll try to overview the whole menu system 
and see can I simplify it.

KevinOConnor commented:
> This is pure Jinja2, it's how jinja is controlling whitespaces.

Okay, thanks.  I was not aware of that.  My reaction is similar though - if would be great if we could come up with a way that the config was not sensitive to extra newlines so that this less common Jinja2 syntax was not needed.  (If the config isn't actually sensitive to newlines, then I'd recommend using the regular `{% ... %}` in the menu configuration file.)

> I have your expectations about the menu system. I'll try to overview the whole menu system
and see can I simplify it.

Perhaps it would help if we could discuss what we think the config should generally look like.  Perhaps mock up a menu.cfg file and/or example-menu.cfg file.  For example, I'm thinking something like:
```
# Common parameters available for all menu config sections.
#[menu some_name]
#type:
#   One of command, input, or list.
#name:
#   Name of menu item - evaluated as a template.
#enable:
#   Template that evaluates to True or False.

#[menu some_command]
#type: command
#name: 
#enable:
#gcode:
#   Code to run on selection.  Evaluated as a template.

#[menu some_input]
#type: input
#name: 
#enable:
#input:
#   Initial value to use when editing - evaluated as a template.
#input_min:
#   Minimum value of range - evaluated as a template.  Default -99999.
#input_max:
#   Maximum value of range - evaluated as a template.  Default 99999.
#input_step:
#   Editing step - evaluated as a template.
#gcode:
#   Code to run on final selection.  Evaluated as a template.
#   Final selection available as {menu.input}.

#[menu some_list]
#type: list
#   Creating a list type allows for menu commands and items to be grouped
#   together in a scrollable list.  Add to the list by creating menu configurations
#   using "some_list" as a prefix - for example: [menu some_list some_entry_in_the_list]
#name: 
#enable:
```

I don't know all the complexities of the menu implementation, so I'm sure I've overlooking some important features.  What other critical features are missing?

Thanks,
-Kevin

mcmatrix commented:
Hi @KevinOConnor 

It took some time but now modified and simplified menu code is in this PR.
It's just for a quick review. I still need to do testing.

Thanks for the menu config example. It's how new menu running at the moment.

mcmatrix commented:
Pushed some updates & adjustments.
Some new menus and a new `text` element.
Made menu for manual calibration (can be improved).
`Setup->Calibration->Delta cal. man`

I thought it's nice to introduce a new scrollable text item for help and other texts
we want to show inside menu.
Please see `Readme 1st` under `Delta cal. man` Click will enter and leave.

mcmatrix commented:
Hi, made some more updates.
I extended a bit text element (new name 'card') capabilities.
Now, the `title` attribute can be used and `scrollbar` can be enabled/disabled.
The purpose of this `card` element is not to replace screens. The idea behind this simple text element is just to add more ways to present info.
To make it more interactive I added a new g-code for menu
`MENU RUN=script NAME=script_name`
This allows triggering topmost container scripts via g-code.
One usage is for menus like this.
```ini
[menu __wait_for_probe_start_card1]
type: card
name: Wait for start
title: Probe
scrollbar: False
script_close:
    {menu.back(True, False)}
script_render:
    " Please wait..."
    " Moving Z:{'%03.2f' % printer.toolhead.position.z}"

[menu __main __setup __calib __delta_calib_man __start]
type: command
name: Start probing
script_press:
    G28
    DELTA_CALIBRATE METHOD=manual
    {script.push_card('__wait_for_probe_start_card1')}
    M400
    MENU RUN=script NAME=close
```
The rendered gcode will be queued.
The script will push `__wait_for_probe_start_card1` to stack (visible for end-user)
It will show user nice waiting text until the printer will home and move the probe to the right height.
`M400` will wait until this is done and then will execute MENU g-code that will call script `close`.
The card will be closed and user will land back to previous menu.

KevinOConnor commented:
Thanks.  I just wanted to let you know that I'm aware of these changes and will take a look at them.  Unfortunately, I have not been able to get time to do that yet.

-Kevin

mcmatrix commented:
I'll resolve the existing conflicts and will drop 'card' item at the moment.
Will update the PR when done.

mcmatrix commented:
I'll fix these conflicts.

mcmatrix commented:
Hi @KevinOConnor 
The updated menu code is in this PR.
Because of a long history of the 'menu upgrades', the menu.py code itself is heavily rewritten.
Sorry for the large diffs but I hope that it's not a problem.

mcmatrix commented:
Hi @KevinOConnor 
Force pushed a fix for the previous code.
Both the `menu.py` and `menu.cfg` are heavily rewritten (huge diffs), everywhere else the changes are minimal.
Please check and let me know about the updated menu.
If something is unclear and concerns, we can discuss.

KevinOConnor commented:
Hi @mcmatrix .

Thanks.  I took a look at the latest version of this PR.  I didn't look at the code changes to menu.py, but I did look through the updated documentation and the new menu.cfg.

The new menu.cfg as a whole looks good to me.  Most of the entries use the basic config settings (type, enable, name, input, input_min, input_max, input_step, gcode) and use basic jinja2 accessors (eg, printer.xxx and menu.input).

However, the new menu documentation (along with a couple of new entries in the menu) introduce several new features that are infrequently used in menu.cfg.  As a whole, my high-level feedback on these additions is similar to my past comments.  My observation on the current menu system is that it is very powerful, but users are struggling to understand and customize it.  If we're changing things up, I'd prefer the new menu system to be as simple as it can.  If after several months of user feedback we find the simple menu system is too burdensome, then I'd look to add additional features at that point.

I'll follow up with some specific comments on the features I'm unsure of.

Thanks again,
-Kevin

KevinOConnor commented on config/example-menu.cfg:
```diff
@@ -3,153 +3,116 @@
-#   input - same like 'command' but has value changing capabilities
-#   list - menu element container, with entry and exit gcode triggers
-#   vsdcard - same as 'list' but will append files from virtual sdcard
+# Menu has default variables defined in menu.cfg
```
I don't think we should define default variables in menu.cfg nor have menu specific variables in general.  All of the config parameters are currently available via the `printer.configfile` accessor.  Ultimately, we don't want users to have to configure limits twice (once for the printer and once for the menu).

For the short-term, I'd leave the hardcoded values in place as they are today.  After merging the new menu code, we can then look to see what values can be pulled from the config, and what values need to be added to various `get_status()` methods.  That way users don't have to configure the printer twice.

KevinOConnor commented on config/example-menu.cfg:
```diff
@@ -3,153 +3,116 @@
-#   vsdcard - same as 'list' but will append files from virtual sdcard
+# Menu has default variables defined in menu.cfg
+# For example, accessible as printer.menu.default.hotend_target_max
+# Variables can be modified in realtime by using 'Setup -> Edit defaults' menu
+# Following template actions can be used:
+#   printer.menu.action_set_default(name, value) - will set <value> for default variable <name>
+#   printer.menu.action_reset_defaults() - will reset saved user default values to the factory defaults
+# Default variables are accessible from the script as {printer.menu.default.default_variable_name}
+# Don't forget to save your config by using 'Setup -> Save config' after modifing defaults.
+#
+# Menu also has an option for temporary runtime variables.
+# Following template action can be used:
+#   printer.menu.action_set_runtime(name, value) - will set <value> for runtime variable <name>
+# Runtime variables are accessible from the script as {printer.menu.runtime.runtime_variable_name}
+# The runtime variables are not stored between restarts.
```
I don't think we should add any menu actions or allow any menu specific runtime configuration.  If a user needs to add something with state then they can add custom menu items that refer to their gcode_macro variables.

KevinOConnor commented on config/example-menu.cfg:
```diff
@@ -3,153 +3,116 @@
+#   menu.ns - element namespace
+#   menu.input - input value, only available in input script context
+# List of actions for menu element:
+#   menu.back(force, update) - will execute menu back command, optional boolean parameters <force> and <update>.
+#                      When <force> is set True then it will also stop editing. Default value is False
+#                      When <update> is set False then parent container items are not updated. Default value is True
+#   menu.exit(force) - will execute menu exit command, optional boolean parameter <force> default value False
+#                      When <force> is set True then it will also stop editing. Default value is False
+#   menu.send_event(name, event, [<arg1> [,<arg2> ...]]) - will emit event 'menu:name:event' with args to klippers event system.
+#                      Event first argument is always source element instance
+#   menu.script_by_name(name, from) - other scripts of <from> item, can be included or imported in jinja2.
+#                                     when <from> is not given then default is '.' itself and '..' is parent.
+# Script attributes and methods:
+#   script.event - name of the event that triggered the script, available in script context
+#   script.log_gcode() - will enable script gcode output logging
```
Similar comment here - I don't think we should add any menu specific actions.

I can see a use for "scripts".  But, at least for now, I'd limit that to the `render()` capability that is currently in display.py.  That way, interested users only need to learn one Klipper Jinja2 script system.

KevinOConnor commented on config/example-menu.cfg:
```diff
@@ -3,153 +3,116 @@
+#   This attribute accepts True or False. When this is enabled then
+#   names longer than display size are scrolled back and forth.
+#   The default is False. This parameter is optional.
+#script_<name>:
+#   Custom script - Evaluated as a template.
```
As before, I don't think the menu should add any additional scripting capabilities.  At least for now.

KevinOConnor commented on config/example-menu.cfg:
```diff
@@ -3,153 +3,116 @@
 #name:
+#enable:
 #cursor:
-#width:
 #scroll:
+#show_title:
+#   Show container name next to back .. element.
+#   The default is True. This parameter is optional.
+#leave_gcode:
+#enter_gcode:
+#   Script to run on entering to list and leaving from list.
+#   Evaluated as a template.
```
I don't think we should implement any of these config options.  I don't think any are used in the menu.cfg currently, and I fear adding these rarely used options makes customizing the menu as a whole harder for users.

KevinOConnor commented on config/example-menu.cfg:
```diff
@@ -3,153 +3,116 @@
 #input_step2:
-#   This is optional attribute for input.
-#   It accepts positive integer or float value. Will determine fast rate
-#   increment and decrement steps for edit value.
+#   Editing step for fast rate - Must be a positive integer or float value.
 #   The default is 0 (input_step will be used instead)
```
FWIW, I'd remove `input_step2`.  I think a simple heuristic would be better.  For example, the code could check if `(input_max - input_min) / input_step > 100` and then enable an `internal_input_step2 = 10 * input_step`.

KevinOConnor commented on config/example-menu.cfg:
```diff
@@ -3,153 +3,116 @@
-#   In edit mode this will be triggered after click button long press (>0.8sec).
-#   The default is empty. This parameter is optional.
+#cursor:
 #reverse:
```
FYI, "reverse" is another option that I don't think should be defined.

mcmatrix commented:
Thank you for the feedback.
I'll check these comments and make necessary adjustments.

mcmatrix commented:
Hi @KevinOConnor 
All adjustments have been made, leaving only what was needed for the menu system.

The menu.cfg has hardcoded values. The input element `input_min` and `input_max` can take script as value.
When it's known from where these constants can be taken from `printer.configfile` then the default menu.cfg 
can be adjusted.

The input step2 is also removed. Implemented your idea `(input_max - input_min) / input_step > 100 and then step is 10 * input_step` as internal step2.

KevinOConnor commented:
Thanks.  My high-level feedback is that it looks good to me.

I saw a couple of minor things:
1. The change to gcode_macro.py doesn't look like it is necessary and I'd prefer not to introduce "property" tags to that class.  (The current Klipper convention is to not access member variables of classes not defined in the same python code file - the property tag makes reviewing for that harder.)
2. I saw a couple of places in the new menu.cfg where `gcode.move_xpos` was changed to `{printer.toolhead.position.x}`, but that isn't correct.  It should change to: `{printer.gcode.gcode_position.x}` .  Basically, if a coordinate is to be passed to a `G1` command then it has to be relative to the "gcode origin" - which means using `printer.gcode.gcode_position`.
3. I saw a couple of places in the new menu.cfg where `G1 E50` was replaced with `G1 E100`.  However, that change doesn't look correct as the default maximum extrusion length is 50.
4. FWIW, I'd also disable the `cursor` config item (at least for now) as it appears to be infrequently used.

Given that this change will break config files of users that have customized the menu, I'm inclined to merge this at the end of this week.  That will give a few days for additional comments to be raised, as well as a few days to get notice out.

Thanks again,
-Kevin

mcmatrix commented:
Thank you for the feedback.
I can make the adjustments before the end of this week (probably on weekend)

mcmatrix commented:
Hi @KevinOConnor

I did all changes you specified.

- Additionally added `printer.configfile.config.extruder.max_temp`, `printer.configfile.config.extruder1.max_temp` and 
`printer.configfile.config.heater_bed.max_temp` into default config.

- Added `SAVE_GCODE_STATE/RESTORE_GCODE_STATE` where `M83/84` or `G90/91` is used in default config.

- Split filament loading/unloading into fast and slow commands.

Please check is that fine?

KevinOConnor commented:
Thanks.  Looks good to me.

FYI, I did notice that `MOVE=1` was incorrectly dropped from the `SET_GCODE_OFFSET Z=...` menu item.

-Kevin

mcmatrix commented:
Thanks for noticing that missing part.
Added it.

KevinOConnor commented:
Thanks.

-Kevin
